pkg/workspace/resource/node.go

```diff
@@ -0,0 +1,297 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resource
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kaitov1beta1 "github.com/kaito-project/kaito/api/v1beta1"
+	"github.com/kaito-project/kaito/pkg/utils"
+	"github.com/kaito-project/kaito/pkg/utils/nodeclaim"
+	"github.com/kaito-project/kaito/pkg/utils/resources"
+	"github.com/kaito-project/kaito/pkg/utils/workspace"
+)
+
+type NodeManager struct {
+	client.Client
+}
+
+func NewNodeManager(c client.Client) *NodeManager {
+	return &NodeManager{
+		Client: c,
+	}
+}
+
+func (c *NodeManager) EnsureNodeResource(ctx context.Context, wObj *kaitov1beta1.Workspace) (bool, error) {
+	// ensure Nvidia device plugins are ready for the workspace when instance type is known.
+	knownGPUConfig, _ := utils.GetGPUConfigBySKU(wObj.Resource.InstanceType)
+	if knownGPUConfig != nil {
+		devicePluginsReady, err := c.ensureNvidiaDevicePluginsReady(ctx, wObj)
+		if !devicePluginsReady || err != nil {
+			return devicePluginsReady, err
+		}
+	}
+
+	// update the workspace status condition and nodes to indicate all resources are ready
+	if err := c.updateWorkspaceStatusIfNeeded(ctx, wObj); err != nil {
+		klog.ErrorS(err, "failed to update workspace status", "workspace", klog.KObj(wObj))
+		return false, err
+	}
+
+	klog.InfoS("All resources are ready for workspace", "workspace", klog.KObj(wObj))
+	return true, nil
+}
+
+// ensureNvidiaDevicePluginsReady ensures that NVIDIA device plugins are ready on all nodes for the workspace
+func (c *NodeManager) ensureNvidiaDevicePluginsReady(ctx context.Context, wObj *kaitov1beta1.Workspace) (bool, error) {
+	requiredProvisionedNodeCount, err := nodeclaim.GetRequiredNodeClaimsCount(ctx, c.Client, wObj)
+	if err != nil {
+		return false, fmt.Errorf("failed to get required node claims count: %w", err)
+	}
+
+	nodes, err := c.getReadyNodesFromNodeClaims(ctx, wObj)
+	if err != nil {
+		if updateErr := workspace.UpdateStatusConditionIfNotMatch(ctx, c.Client, wObj, kaitov1beta1.ConditionTypeResourceStatus, metav1.ConditionFalse,
+			"NodeListError", fmt.Sprintf("Failed to get nodes for workspace: %v", err)); updateErr != nil {
+			klog.ErrorS(updateErr, "failed to update resource status condition", "workspace", klog.KObj(wObj))
+		}
+		return false, fmt.Errorf("failed to get nodes for workspace: %w", err)
+	}
+
+	if len(nodes) != requiredProvisionedNodeCount {
```

Comment:
When scaling down, len(nodes) may larger than requiredProvisionedNodeCount . what's the expected behavior?